### ISSUE TITLE
AzureProvisoner output saves connection strings to user secrets even if it did nothing

### ISSUE THREAD

**Issue by davidfowl** [dotnet/aspire#3298]
This is what happens when you use an azure resource and do nothing else:

```
Building...
info: Aspire.Hosting.Azure.AzureProvisioner[0]
      Using DefaultAzureCredential for provisioning. This may not work in all environments. See https://aka.ms/azsdk/net/identity/default-azure-credential for more information.
info: Aspire.Hosting.Azure.AzureProvisioner[0]
      Azure resource connection strings saved to user secrets.
```

The first log should only show up if we are going to provision.
The second log should only show up if we have indeed saved connection string to user secrets.

**Comment by scabana** [comment]
I don't want to open another issue about `AzureProvisioner` and user secrets. I have a feeling a provisioner api would help with those issues.

The way it's coded, anyone taking the `AzureProvisioner` as an example to have a custom provisioner could encounter a concurrency issue updating the user secrets. As the provisioner loads the user secrets once and then saves it at the end of the work (which is launched in `Task.Run`). Another provisioner that would do the same could end up losing / overriding modified user secrets from another provisioner.

**Comment by davidfowl** [comment]
Yep, it's problematic. We'd need to expose a first class API. That wouldn't help if modifications were made outside of the app host though.

**Comment by joperezr** [comment]
As part of triage cleanup ahead of the 13.3 release, this issue is being moved to the Backlog milestone. If this is already fixed or stale it should be closed.

**Comment by mauri0686** [comment]
I checked this on current main (76e76c5) and I believe it's fixed. The "connection strings saved to user secrets" message no longer exists. A resource served from cached state now logs "Using connection information stored in user secrets for {resourceName}." (`AzureProvisioningController.cs:3963`) and nothing is written back. The credential message (`DefaultTokenCredentialProvider.cs:79`) only appears when a provisioning context is created, and that context is now lazy (`AzureProvisioningController.cs:3893`), so a fully cached run never authenticates.

I couldn't find a test asserting that a fully cached run creates no provisioning context. I'm happy to send a small test-only PR to lock that in; otherwise this can probably be closed.


> [!NOTE]
> Drafted with AI assistance (Claude Code) and reviewed by me.
